File: projeto/modelo/NewtonCotes.cpp
#include "NewtonCotes.h"

NewtonCotes::NewtonCotes(){
	filosofia = 1;
	grau = 1;
}

void NewtonCotes::setFilosofia(int filosofia){
	this->filosofia = filosofia;
}

void NewtonCotes::setGrau(int grau){
	this->grau = grau;
}


float NewtonCotes::calcular(Funcao *funcao, float Xi, float Xf){
	float resultado;
	if (filosofia) { //1 PARA A FILOSOFIA FECHADA 
		switch(grau) {
			case 1: // GRAU 1
				resultado = ((Xf - Xi)/2)*(funcao->calcular(Xi) + funcao->calcular(Xf));
			break;

		case 2: // GRAU 2
				resultado = ((Xf - Xi)/9)*(funcao->calcular(Xi) + 4*funcao->calcular((Xi+Xf)/2) + funcao->calcular(Xf));
			break;

			case 3: // GRAU 3
				resultado = 0;
			break;

			case 4: // GRAU 4
				resultado = 0;
			break;
		}
	} else { //0 PARA FILOSOFIA ABERTA
		switch(grau) {
			case 1: // GRAU 1
				resultado = 0;
			break;

			case 2: // GRAU 2
				resultado = 0;
			break;

			case 3: // GRAU 3
				resultado = 0;
			break;

			case 4: // GRAU 4
				resultado = 0;
			break;
		}
	}
	return resultado;
}

File: Integrais/src/main/scala/Modelo/Integrais/NewtonCotes.scala
package Modelo.Integrais
import Modelo.Funcao.Funcao

/**
  * Implementacao da integral Newton-Cotes
  */
object NewtonCotes extends Integral {

  /** Metodo para calcular a integral usando Newton-Cotes
    *
    * @param expressao funcao generica
    * @param valor_inicial 'x' inicial
    * @param valor_final 'x' final
    * @param grau grau do polinomio de integracao
    * @param filosofia filosofia desejada para o calculo. True = Forward e False = Backward
    * @return o valor da integral
    */
  override def apply(expressao: Funcao, valor_inicial: Double, valor_final: Double , grau:Int = 1 , filosofia: Boolean = true): Double ={
    if(filosofia){

      grau match {

        case 1 => ((valor_final - valor_inicial)/2) * expressao(valor_inicial) + expressao(valor_final)

        case 2 => ((valor_final - valor_inicial)/9) * expressao(valor_inicial) + 4*expressao((valor_inicial+valor_final)/2)

      }
    }else{

      grau match {

        case 1 => 0

      }
    }
  }
}
